### ISSUE TITLE
motion correction: avoid loading entire movie into memory

### ISSUE THREAD

**Issue by tbenst** [flatironinstitute/CaImAn#749]
For better support, please use the template below to submit your issue. When your issue gets resolved please remember to close it.

Sometimes errors while running CNMF occur during parallel processing which prevents the log to provide a meaningful error message. Please reproduce your error with setting `dview=None`.

If you need to upgrade CaImAn follow the instructions given in the [documentation](https://caiman.readthedocs.io/en/master/Installation.html#upgrading).

- **Tell us a bit about your setup**: 
 1. Operating system (Linux/macOS/Windows):   linux

 2. Python version (3.x): 3.7
 
 3. Working environment (Python IDE/Jupyter Notebook/other):
 
 4. Which of the demo scripts you're using for your analysis (if applicable): 3d motion correct
 
 5. CaImAn version*: 1.8.4
 
 6. CaImAn installation process (`pip install .`/`pip install -e .`/conda): source

- **Describe the issue that you are experiencing**

I have a 155GB 3D file (600 x 201 x 672 x 672, T x Z x H x W, float32, compressed modestly). My computer has 128GB RAM + 16GB swap. My file is NWB with hdf5 backend. CaImAn tries to read the entire file into memory, which promply gets killed by the kernel.

Could I request that CaImAn add support for files larger than can fit into RAM? This file is only 1 minute of recording, so using more RAM isn't a scaleable solution.

I know that `cm.load` already supports `in_memory=False`, and it seems that the code already has some thought put in for [batch processing](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L2745).

I know I need to manually set `min_mov` to [avoid blowing up mem usage](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L235).

When I run the code with logging set to DEBUG, I see that `DEBUG:root:Number of Splits: 6`, and subsequently I see `Processing: frames: [<100 numbers here>]` six times. After seeing this, CaImAn loads everything into ram.

This is odd, as my mc_dict from `opts = params.CNMFParams(params_dict=mc_dict)` has `'splits_els': 20,  'splits_rig': 20, 'num_frames_split': 30,`, thus I expect 20 splits of 30 frames. I have `dview=None` and as expected I don't see [this message](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L3030).

I never see [count](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L2925) appear, so my blow-up happens before that.

there's [`cm.load`](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L2918), but even 100 frames should only be 100*201*672^2*4/1024**3=33.8GiB.

That leads me to https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L2920, but still should only be at 70 GiB or so.

In summary, 
1) num_frames_split / splits_rig seems to be ignored, which is a bug.
2) memory blows up, and I can't tell where...

**Comment by epnev** [comment]
@tbenst 
1. I would check whether the variables `dims, T, splits` have the right values after this [point](https://github.com/flatironinstitute/CaImAn/blob/8f2722a1e5523df104ea95899cb35e3efa269cf7/caiman/motion_correction.py#L2978) 
2. My suspicion is that the outage occurs at [this point](https://github.com/flatironinstitute/CaImAn/blob/6c5c3b6117b71b6e8b44f62fc26fd3b3d914de12/caiman/motion_correction.py#L2918) and is possibly related to your first point. 
Can you create a print/log message right above this point that captures the value of `idx`?

**Comment by tbenst** [comment]
1. No, they do not have the right values at that point as indicated by `DEBUG:root:Number of Splits: 6` from https://github.com/flatironinstitute/CaImAn/blob/8f2722a1e5523df104ea95899cb35e3efa269cf7/caiman/motion_correction.py#L2981

2. good idea will report back when I have the chance, have a lab meeting coming up so sticking to 2D for now...

**Comment by SabriQ** [comment]
motion correction is get stuck in cluster (SLURM,1TB RAM) when motion correction. All I can see is its log get stuck in somewhere of motion correction , as if paused videostream.
I guess it's not due to the script all the RAM, because there is no any error output and I can still see normal memory usage.
whether it is due to the tasks running on the same node as mine?
I'm not sure whether it's an issue, if yes ,I can open a new issue out of here.

**Comment by j-friedrich** [comment]
Long period of inactivity, but the issue has now been addressed in the dev branch (commit 7a7f4021609406d631c6018b9e876a1225e5a8a5)
